colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -0,0 +1,848 @@
+        """
+        # Load input ids, attention mask and labels
+        # micro_batch = self.load_micro_batch(model_chunk_id=model_chunk_id)
+
+        # for the first stage, input_obj is None
+        # for other stages, input_obj is the output of the previous/next stage containing hidden_states etc.
+        # Only attention_mask from micro_batch is used
+
+        with self.stage_manager.switch_model_chunk_id(model_chunk_id):
+            # fwd calculate
+            output_obj = model_chunk[model_chunk_id](input_obj)
+            # last layer in model
+            if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+                loss = criterion(output_obj) / self.num_microbatch
+                if accum_loss is not None:
+                    accum_loss.add_(loss.detach())
+                if outputs is not None:
+                    outputs.append(tree_map(detach, output_obj))
+                return loss
+            else:
+                return output_obj
+
+    def backward_b_step(
+        self,
+        model_chunk: Union[ModuleList, Module],
+        model_chunk_id: int,
+        optimizer: OptimizerWrapper,
+        input_obj: Optional[dict],
+        output_obj: Union[dict, torch.Tensor],
+        output_obj_grad: Optional[dict],
+    ) -> Optional[dict]:
+        """Backward dx step of the pipeline; we calculate "dx = w*dy" here;
+
+        Args:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            optimizer (OptimizerWrapper): Optimizer to update the model
+            input_obj (Optional[dict]): x.
+            output_obj (Union[dict, torch.Tensor]): y.
+            output_obj_grad (dict): dy.
+
+        Returns:
+            Optional[dict]: dx.
+        """
+        # calculate bwd b step ; only dx = w*dy;
+
+        # Retain the grad on the input_obj.
+        tree_map(retain_grad, input_obj)
+
+        if model_chunk_id == 0:
+            # bwd step
+            optimizer.backward_by_grad(
+                tensor=output_obj,
+                grad=output_obj_grad,
+                inputs=input_obj,
+                retain_graph=True,
+            )
+        else:
+            if self.stage_manager.is_first_stage(ignore_chunk=True):
+                # loss backward; output_obj is loss
+                optimizer.backward_by_grad(
+                    tensor=output_obj,
+                    grad=None,
+                    inputs=input_obj,
+                    retain_graph=True,
+                )
+
+            else:
+                # commom bwd step
+                optimizer.backward_by_grad(
+                    tensor=output_obj,
+                    grad=output_obj_grad,
+                    inputs=input_obj,
+                    retain_graph=True,
+                )
```

Comment:
It seems that this branch is the same as if `model_chunk_id == 0`

Reply:
> It seems that this branch is the same as if `model_chunk_id == 0`

simplify in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/4c1f81c68356669af9d3ccd8b3d395c3db97afbb;